### PR TITLE
[Windows] shrink window to fit the screen and avoid a crash

### LINKED ISSUE #22667: Kodi 20 - Crash at Startup on Windows 11 Pro (probably related to screen switching)
### Describe the bug
[I am not a native English speaker]
Here is a clear and concise description of what the problem is:
Kodi crashes (hard crash) at startup. The behaviour can be reproduced. Windows event log entry, stack trace and .dmp files are created.

Trying around and looking at debug logs myself, I believe to have recognized the cause of the bug. It happens on a notebook with Full HD screen, that sometimes is attached to via a docking station to a 4k monitor.
1. Use Kodi with the 4k monitor (starts perfectly here). Use windowed mode (not full screen). However I did not investigate all the conditions. Exit Kodi normally.
2. Pull off the monitor/docking station
3. Start Kodi -> A black Window opens, nothing but black rectangle, closes again. Stack-Trace and Windows event log created
This happened with Kodi 19.4 (when I did not understand yet, that I will be able to successfully start it again, when attaching the monitor) and also with 20.0 under Windows 11 Pro

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

Normal start

## Actual Behavior

Crash

## Possible Fix

No idea

### To Reproduce

Text from above:

1. Use Kodi with the 4k monitor (starts perfectly here). Use windowed mode (not full screen). However I did not investigate all the conditions. Exit Kodi normally.
2. Pull off the monitor/docking station
3. Start Kodi -> A black Window opens, nothing but black rectangle, closes again. Stack-Trace and Windows event log created

### Debuglog

The debuglog can be found here:
https://paste.kodi.tv/zilepaduzu.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem:

no screenshots

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

Stack trace: 
Thread 14496 (process 13400)
# 0
Event log text:
Name der fehlerhaften Anwendung: kodi.exe, Version: 20.0.0.0, Zeitstempel: 0x63c3d036
Name des fehlerhaften Moduls: kodi.exe, Version: 20.0.0.0, Zeitstempel: 0x63c3d036
Ausnahmecode: 0xc0000005
Fehleroffset: 0x000000000083fd51
ID des fehlerhaften Prozesses: 0x0x2220
Startzeit der fehlerhaften Anwendung: 0x0x1D9368560D0B9B4
Pfad der fehlerhaften Anwendung: C:\Program Files\Kodi\kodi.exe
Pfad des fehlerhaften Moduls: C:\Program Files\Kodi\kodi.exe
Berichtskennung: 31d763c7-e90b-4bd5-83e3-1ed85c2c9aac
Vollständiger Name des fehlerhaften Pakets: 
Anwendungs-ID, die relativ zum fehlerhaften Paket ist: 

Provider 

   [ Name]  Application Error 
   [ Guid]  {a0e9b465-b939-57d7-b27d-95d8e925ff57} 
 
   EventID 1000 
 
   Version 0 
 
   Level 2 
 
   Task 100 
 
   Opcode 0 
 
   Keywords 0x8000000000000000 
 
  - TimeCreated 

   [ SystemTime]  2023-02-01T21:37:28.0273857Z 
 
   EventRecordID 14828 
 
   Correlation 
 
  - Execution 

   [ ProcessID]  10324 
   [ ThreadID]  5464 
 
   Channel Application 
 
   Computer LAPTOP-F7JV4U81 
 
  - Security 

   [ UserID]  S-1-5-21-3767424003-1292986099-2155062632-1001 
 

- EventData 

  AppName kodi.exe 
  AppVersion 20.0.0.0 
  AppTimeStamp 63c3d036 
  ModuleName kodi.exe 
  ModuleVersion 20.0.0.0 
  ModuleTimeStamp 63c3d036 
  ExceptionCode c0000005 
  FaultingOffset 000000000083fd51 
  ProcessId 0x2220 
  ProcessCreationTime 0x1d9368560d0b9b4 
  AppPath C:\Program Files\Kodi\kodi.exe 
  ModulePath C:\Program Files\Kodi\kodi.exe 
  IntegratorReportId 31d763c7-e90b-4bd5-83e3-1ed85c2c9aac 
  PackageFullName  
  PackageRelativeAppId  

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] macOS
 - [ x] Windows
 - [ ] Windows UWP

 - Operating system version/name: Windows 10 Pro
 - Kodi version: 20.0 (also 19.4)

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by CrystalP**
## Description

In windowed mode, adjust width and height from settings so that the window fits on the screen.
Take into account occlusions such as the Windows task bar and window Title bar / decorations.

I think this is a low risk approach that could be backported to v20,

A better fix untangling RenderSystem/WindowSystem, refining the new position and size and avoiding changing the settings value until the use changes the size or position will take more time and have more risk - probably fit for v21 only.

## Motivation and context

Kodi crashes in windowed mode when going to a smaller screen.
Steps:
1. Start Kodi on external 4k screen, windowed mode
2. Stop Kodi, disconnect external screen
3. Start Kodi on the builtin 1080p screen
4. crash

reported in issue #22667, and probably what is alluded to in some forum threads.

root cause: when the window is much larger than the available screen, Windows doesn't send an initial WM_SIZE message that indirectly intializes the swap chain and other DX resources that CRenderSystemDX::InitRenderSystem() relies on but does not initialize.
The message normally happens as a side-effect of CWinSystemWin32::CreateNewWindow() when the window fits in the screen.

That hidden dependency should not exist.
## How has this been tested?

Windows 10, 2 screens, followed the steps above and more common situations of switching between windowed and fullscreen, resizing of the window.

## What is the effect on users?

Fix a crash on startup in windowed mode after unplugging a larger screen.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/windowing/windows/WinSystemWin32.cpp [hunk 1]
```diff
@@ -1354,3 +1392,28 @@ void CWinSystemWin32::CacheSystemSdrPeakLuminance()
...
+  RECT rcNcArea{};
+  SetRectEmpty(&rcNcArea);
+
+  if (!AdjustWindowRectEx(&rcNcArea, dwStyle, false, dwExStyle))
+  {
+    CLog::LogF(LOGERROR, "AdjustWindowRectEx failed with {}", GetLastError());
+    return RECT();
+  }
+  return rcNcArea;
+}
```
Comment (thexai) [inline xbmc/windowing/windows/WinSystemWin32.cpp:1419]: Missing empty line at end

Comment (CrystalP) [reply xbmc/windowing/windows/WinSystemWin32.cpp]: indeed, fixed

**Review by thexai — APPROVED** [review]

**Comment by axbmcuser** [comment]
@CrystalP 

I suspect this PR is causing/forcing the Kodi window a bit too small in the following scenario/example:

1920x1200 screen, Kodi windowed mode, window size manually set to exactly 1920x1080 
(verified via Kodi system info dialog. Works that way for years.)

Before this PR: 
Pixel perfect window size of 1920x1080 kept and correctly stays that way after restarting Kodi
Kodi 16:9 skins can be used as intended in exact 16:9 format without visual issues

After this PR:
Window size is always forced+reduced to 19**02**x1080 on every Kodi start

Any chance you could have a look again?

I get the purpose of the fix, but imho it should not touch windows which in fact aren't larger than the screen resolution and force them to a "wrong"/not exact window size which causes the skin not to be pixel perfect anymore. 

Thank you! :)

**Comment by CrystalP** [comment]
Yes this PR is the reason. It was done that way because it's unclear how far off screen you can be before Windows stops sending the WM_SIZE Kodi needs for initialization.
Even though you may not have realized it, your settings cause the window's left and right borders to be off screen.
More exploration is needed to find where the limit is to take care of the original issue addressed by the PR.

**Comment by CrystalP** [comment]
@axbmcuser since the fix was merged in Omega a couple days ago, please try the latest v21 nightly to confirm the solution.

**Comment by axbmcuser** [comment]
@CrystalP 
Just made some tests with 323ea7f9 from 2023-03-23:
- I can confirm the described use case is working again with exactly kept window size of 1920x1080
- Made a further test with 1921x1080 - in which case the window size is not kept which may indicates that the PRs intentions are still working and we now should have the best of both worlds.

Thanks for adressing it! 👍

**Pull request merged**